File: hybrid/hybrid-hello/solution/hybrid-hello.c
#include <stdio.h>
#include <mpi.h>
#include <omp.h>

int main(int argc, char *argv[])
{
    int my_id, omp_rank;
    int provided, required=MPI_THREAD_FUNNELED;

    MPI_Init_thread(&argc, &argv, required, &provided);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_id);

#pragma omp parallel private(omp_rank)
    {
        omp_rank = omp_get_thread_num();
        printf("I'm thread %d in process %d\n", omp_rank, my_id);
    }

    if (my_id == 0) {
        printf("\nProvided thread support level: %d\n", provided);
        printf("  %d - MPI_THREAD_SINGLE\n", MPI_THREAD_SINGLE);
        printf("  %d - MPI_THREAD_FUNNELED\n", MPI_THREAD_FUNNELED);
        printf("  %d - MPI_THREAD_SERIALIZED\n", MPI_THREAD_SERIALIZED);
        printf("  %d - MPI_THREAD_MULTIPLE\n", MPI_THREAD_MULTIPLE);
    }
    MPI_Finalize();
    return 0;
}

File: openmp/hybrid-hello/solution/hybrid-hello.f90
program hello
    use omp_lib
    use mpi
    implicit none
    integer :: my_id, tid, rc
    integer :: provided, required=MPI_THREAD_FUNNELED

    call MPI_Init_thread(required, provided, rc)
    !call MPI_Init(rc)
    !call MPI_Query_thread(provided, rc)

    call MPI_Comm_rank(MPI_COMM_WORLD, my_id, rc)

    !$omp parallel private(tid)
    tid = omp_get_thread_num()
    write(*,'(A,I2,A,I2)') "I'm thread ", tid, ' in process ', my_id
    !$omp end parallel

    if (my_id == 0) then
        write(*,*) ''
        write(*,'(A,I1)') 'Provided thread support level: ', provided
        write(*,'(A,I1,A)') '  ', MPI_THREAD_SINGLE, &
                            & ' - MPI_THREAD_SINGLE'
        write(*,'(A,I1,A)') '  ', MPI_THREAD_FUNNELED, &
                            & ' - MPI_THREAD_FUNNELED'
        write(*,'(A,I1,A)') '  ', MPI_THREAD_SERIALIZED, &
                            & ' - MPI_THREAD_SERIALIZED'
        write(*,'(A,I1,A)') '  ', MPI_THREAD_MULTIPLE, &
                            & ' - MPI_THREAD_MULTIPLE'
    end if
    call MPI_Finalize(rc)
end program hello
